dev-docs/RFCs/v8.0/ui-module-rfc.md

```diff
@@ -0,0 +1,137 @@
+  setProps: (props: PropsT) => void;
+
+  /// Event hooks
+  onViewStateChange?: () => void;
+  onRedraw?: () => void;
+  onHover?: (info) => void;
+  onClick?: (info) => void;
+}
+```
+
+2. Add a new module `@deck.gl/ui` which exports common UI components that implement the `UIControl` interface.
+
+3. Add the following APIs to the `Deck` class:
+
+```js
+/// Prop type for React / declarative usage
+new Deck({
+  uiControls: UIControl[]
+})
+
+/// Imperative methods for vanilla JS usage
+addUIControl(control: UIControl): void
+removeUIControl(control: UIControl): void
+```
+
+Users will use this new module like the following:
+
+```js
+import {NavigationControl, FullscreenControl} from '@deck.gl/ui'
+
+new Deck({
+  ...
+  uiControls: [
+    new NavigationControl(),
+    new FullscreenControl()
+  ]
+})
+```
+
+## Implementation details and concerns
+
+### Terminology
+
+React and Mapbox users use "controls" to refer to UI components. However, we already call an input handler "controller" in the deck.gl context. I have seen users not differentiating the two in issues and discussions. I am concerned that introducing another concept called "control" is going to add to the confusion.
+
+### React
+
+All components in `@deck.gl/ui` should work without React. `@deck.gl/react` can offer thin wrappers that work with the `DeckGL` component:
+
+```jsx
+import {NavigationControl, FullscreenControl} from '@deck.gl/react'
+
+<DeckGL>
+  <NavigationControl />
+  <FullscreenControl />
+</DeckGL>
+```
+
+### Performance impact
+
+A new `UIControlManager` will be added to `Deck` alongside `ViewManager` `EffectManager` etc. It will handle calling the lifecycle and event hooks of registered controls. For existing apps, where no UI controls are used, the performance impact is projected to be neglegible.
+
+I would like to call out the expected behavior when the `uiControls` prop shallowly changes. Following existing behavior pattern of `layers`, `views` and `effects`, upon `setProps`:
+
+- `UIControlManager` deeply compares the elements of `uiControls`' old and new values. Two controls are be the "same" if they share the same constructor and the same id.
+- If a control only exists in the new array, call `onAdd`
+- If a control only exists in the old array, call `onRemove`
+- If a control exists in both old and new arrays, transfer the new props to the old instance, instead of adopting the new instance
+
+This would avoid the performance hit by constantly adding/removing controls when using React.
+
+### CSS packaging
+
+react-map-gl (inherited from mapbox-gl) requires a separate CSS stylesheet be included in the page's head. deck.gl traditionally uses CSS-in-JS (see [getTooltip](/docs/api-reference/core/deck.md)). While I like the convenience of everything-in-one-bundle approach, the new UI module will have much more complex DOM structures, and it will be harder to design an API that allows users to customize the look and feel of their UI.
```

Comment:
Separate to styling is the question of layout. While often this is achieved via CSS, this can be error-prone and hard for users, especially if it means deploying a separate CSS file.

A common use case will be something like "put the compass in the top left, zoom controls on the bottom left, and the other controls along the right side etc". Would it be worth adding a parameter to the controls to specify the placement and have apply the correct CSS positioning, e.g.

```js
new Deck({
  ...
  uiControls: [
    new NavigationControl({placement: 'bottom-left'}),
    new FullscreenControl({placement: 'top-right'})
  ]
})
```